File: backend/app/schemas/message.py
"""
schemas/message.py — Schemas Pydantic para el modelo Message

Los schemas de Pydantic definen la "forma" del JSON que entra y sale de la API.
Son la barrera de seguridad real del servidor (a diferencia de los types de
TypeScript que solo guían al desarrollador en el editor).

Flujo de lectura recomendado para entender este módulo:
  1. models/message.py     → estructura de la tabla en PostgreSQL
  2. schemas/message.py    → forma del JSON de entrada/salida (este archivo)
  3. routers/message.py    → endpoints que usan estos schemas
"""

import datetime
from typing import Optional

from pydantic import BaseModel, Field


class MessageCreate(BaseModel):
    """
    Schema de ENTRADA para crear un nuevo mensaje en un ticket.

    Solo se necesita el contenido porque los demás campos (ticket_id, author_id,
    is_system, timestamps) los infiere el servidor desde el contexto del request:
      - ticket_id  → viene de la URL del endpoint (/tickets/{ticket_id}/...)
      - author_id  → viene del JWT en la cookie (usuario autenticado)
      - is_system  → siempre False para mensajes creados por humanos vía este schema
      - created_at → la BD lo genera automáticamente

    Attributes:
        content: El texto del mensaje. Mínimo 1 caracter para evitar mensajes vacíos.
    """

    content: str = Field(..., min_length=1, description="Contenido del mensaje")


class MessageAuthor(BaseModel):
    """
    Schema anidado que representa al autor de un mensaje.

    Diseño: El anidamiento directo de la entidad `Author` evita el problema de
    N+1 peticiones HTTP desde el frontend, proporcionando todos los datos
    necesarios para la renderización de la interfaz (ej. nombre e email)
    en una sola respuesta inicial.

    Attributes:
        id:        ID del usuario autor.
        email:     Email del usuario (usado como fallback si full_name es None).
        full_name: Nombre completo del usuario. Puede ser None si el usuario
                   no lo ha configurado en su perfil.
    """

    id: int
    email: str
    full_name: Optional[str] = None

    # from_attributes=True le dice a Pydantic que puede construir este schema
    # desde un objeto SQLAlchemy (message.author) además de desde un dict.
    model_config = {"from_attributes": True}


class MessageRead(BaseModel):
    """
    Schema de SALIDA para devolver un mensaje con todos sus datos al frontend.

    Incluye el objeto 'author' anidado con nombre y email para que el frontend
    pueda mostrar el autor sin requests adicionales.

    Integración con SQLAlchemy:
    El campo `author` utiliza `from_attributes=True` para permitir a Pydantic
    extraer los datos de la relación Mapped["User"] del modelo SQLAlchemy
    (obtenidos eficientemente vía `joinedload`). Dado que los nombres de los
    atributos coinciden, no se requiere configuración manual adicional.

    Attributes:
        id:         ID autoincremental del mensaje en la BD.
        content:    Texto del mensaje.
        is_system:  True si fue generado automáticamente por el sistema
                    (ej. "Ticket asignado a X"). False si lo escribió un humano.
        ticket_id:  ID del ticket al que pertenece el mensaje.
        author_id:  ID numérico del autor (útil para comparar con currentUser.id en el frontend).
        author:     Objeto anidado con id, email y full_name del autor.
                    Pydantic lo lee directamente de message.author (relación SQLAlchemy).
        created_at: Timestamp de creación, para ordenar y mostrar la hora del mensaje.
    """

    id: int
    content: str
    is_system: bool
    ticket_id: int
    author_id: int
    author: Optional[MessageAuthor] = None
    created_at: datetime.datetime

    model_config = {"from_attributes": True}

File: frontend/src/types/message.ts
/**
 * types/message.ts — Tipos TypeScript para el sistema de mensajes
 *
 * Estos tipos son el espejo del schema Pydantic `MessageRead` del backend.
 * ¿Para qué sirven si Pydantic ya valida en el servidor?
 *   - Guían al desarrollador mientras programa (autocompletado en VS Code).
 *   - Detectan errores de tipado en tiempo de compilación (no en producción).
 *   - Son eliminados por TypeScript al compilar — no existen en el navegador.
 *
 * Nota sobre `created_at: string` (no `Date`):
 *   Python serializa datetime a ISO 8601 string ("2026-05-02T14:35:00.000Z")
 *   al viajar por HTTP/WebSocket. TypeScript recibe ese string. La conversión
 *   a objeto Date se hace en el componente cuando se quiere formatear la hora.
 *
 * Flujo de lectura:
 *   backend/schemas/message.py → este archivo → services/messageService.ts → ChatBox.tsx
 */

/**
 * Datos del autor de un mensaje.
 *
 * Espejo del schema `MessageAuthor` de Pydantic.
 * Se usa para mostrar el nombre completo del autor (estilo Teams).
 *
 * @property id        - ID numérico del usuario en la BD.
 * @property email     - Email del usuario. Usado como fallback si full_name es null.
 * @property full_name - Nombre completo. Puede ser null si el usuario no lo configuró.
 */
export interface MessageAuthor {
  id: number;
  email: string;
  full_name: string | null;
}

/**
 * Representa un mensaje completo con los datos de su autor.
 *
 * Espejo del schema `MessageRead` de Pydantic.
 *
 * @property id        - ID autoincremental del mensaje.
 * @property content   - Texto del mensaje.
 * @property is_system - true = generado por el sistema (log de auditoría, ej. "Ticket asignado a X").
 *                       false = escrito por un humano (mensaje de chat normal).
 * @property ticket_id - ID del ticket al que pertenece.
 * @property author_id - ID del autor (útil para comparar con currentUser.id y saber si el mensaje es "mío").
 * @property author    - Objeto con datos completos del autor para mostrar en la UI.
 * @property created_at - Timestamp ISO 8601 de cuándo se creó el mensaje.
 */
export interface MessageRead {
  id: number;
  content: string;
  is_system: boolean;
  ticket_id: number;
  author_id: number;
  author: MessageAuthor;
  created_at: string;
}

/**
 * Eventos que el servidor puede enviar por WebSocket.
 *
 * El campo `type` actúa como discriminador para que el frontend sepa
 * cómo manejar cada evento. Hay tres tipos posibles:
 *
 * - "chat_message":      Nuevo mensaje de chat (humano). Incluye todos los campos de MessageRead.
 * - "ticket_assigned":   Alguien se asignó el ticket. Incluye datos del nuevo responsable.
 * - "ticket_unassigned": El ticket fue devuelto a la bandeja sin responsable.
 */
export type ChatMessageEvent = MessageRead & { type: "chat_message" };

export interface TicketAssignedEvent {
  type: "ticket_assigned";
  assigned_to: MessageAuthor;
}

export interface TicketUnassignedEvent {
  type: "ticket_unassigned";
}

/** Unión de todos los eventos posibles que puede enviar el servidor por WebSocket */
export type WebSocketEvent = ChatMessageEvent | TicketAssignedEvent | TicketUnassignedEvent;
